### ISSUE TITLE
Format string reentrency

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#2244]
With b94ee60b07b4c7e204e5f9e06aeba462fc8277f6, it is now possible to have format strings in format strings:  `'Hello ${name+'$age years'} old'`

However there seems to be an issue if the inner string has spaces around it:

`'${    '${x}'}'` does not work (the inner string is not considered a format string)

**Comment by ncannasse** [comment]
That's actually quite complex and requires a lot of testing. Any volunteer to take over this task ? @Simn   @nadako  ?

**Comment by Simn** [comment]
I think we should first find a better way to communicate the quoted status (" or ') for strings, which would make this a Haxe 4 issue.

**Comment by ncannasse** [comment]
you mean inside the string itself ? this should not be required, given that we have all the information needed to correctly lex it

**Comment by Simn** [comment]
> '${ '${x}'}' does not work (the inner string is not considered a format string)

This suggests to me that the information is lost somewhere. I'm not very familiar with the current implementation, but in general inner format strings should be handled without any additional effort via normal expression recursion.

**Comment by ncannasse** [comment]
Yes, I agree, this is normally done directly in lexer.mll (string2, code_string) which should unquote things as they should, but doesn't

**Comment by Simn** [comment]
Don't we need a stack in order to lex nested structures like this? At the moment we only have `buf` which is reset on new strings/comments/regexes, so that really doesn't support any notion of "string in string".

**Comment by Simn** [comment]
Eh, I guess we don't because we dump everything into the same buffer and deal with it while typing.

**Comment by Simn** [comment]
I fixed the spacing issue. As for this: 

> Another related issue: we should allow any code in ${} , not only values.
> For instance the following does not work ATM:
> '${ var str = ""; str; }' while the following works : '${{ var str = ""; str; }}'

I consider that a fair limitation. If you want a block you have to start a block. Sure it looks a bit funny because interpolation also uses `{}` as delimiters, but I don't see why we should imply blocks here.

> Another last issue is that quoted chars in sub format strings are unquoted.

Any example of that?

**Comment by ncannasse** [comment]
`'${'\''}'` should be ok, but it's not. I think that requires an entirely different way of processing format strings.

**Comment by Aurel300** [comment]
Related issues:

 - #3974
 - #5985
 - #8888 - `'${"\\\\"}'` should produce two backslashes, but only produces one
 - #10287 - `'${foo()blah}'`, `blah` is completely ignored
 - #10463 - `'${'\''}'`

And here is a new one:

```haxe
class Main {
  public static function main():Void {
    var foo = "hello";
    trace('${'${foo}'}'); // outputs hello
    trace('\n${'${foo}'}'); // outputs ${foo}
  }
}
```

**Comment by nadako** [comment]
I'm going to revive my old branch with the "mini-ast" and probably do what is suggested in https://github.com/HaxeFoundation/haxe/issues/9370#issuecomment-621765541.

**Comment by jeremyfa** [comment]
Hitting this, which looks related (just FYI):

Fails:
![image](https://github.com/user-attachments/assets/27ce7d64-2b17-41f6-8f9e-92e9bbdcb3ff)

Accepted by Haxe:
![image](https://github.com/user-attachments/assets/3a84a377-df38-4d16-86b9-0ee55ffb2752)

Each new level need one more escaping:
![image](https://github.com/user-attachments/assets/fda93a74-4c4d-445f-8d04-5e0785ae785d)
